pkg/services/authz/zanzana/store/migration/migrator.go

```diff
@@ -0,0 +1,124 @@
+package migration
+
+import (
+	"embed"
+	"errors"
+	"fmt"
+	"strings"
+
+	"xorm.io/xorm"
+
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+func Run(cfg *setting.Cfg, typ, connStr string, fs embed.FS, path string) error {
+	engine, err := xorm.NewEngine(typ, connStr)
+	if err != nil {
+		return fmt.Errorf("failed to parse database config: %w", err)
+	}
+
+	m := migrator.NewMigrator(engine, cfg)
+	m.AddCreateMigration()
+
+	return RunWithMigrator(m, cfg, fs, path)
+}
+
+func RunWithMigrator(m *migrator.Migrator, cfg *setting.Cfg, fs embed.FS, path string) error {
+	migrations, err := getMigrations(fs, path)
+	if err != nil {
+		return err
+	}
+
+	for _, mig := range migrations {
+		m.AddMigration(mig.name, mig.migration)
+	}
+
+	sec := cfg.Raw.Section("database")
+	return m.Start(
+		sec.Key("migration_locking").MustBool(true),
+		sec.Key("locking_attempt_timeout_sec").MustInt(),
+	)
+}
+
+type migration struct {
+	name      string
+	migration migrator.Migration
+}
+
+func getMigrations(fs embed.FS, path string) ([]migration, error) {
+	entries, err := fs.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read migration dir: %w", err)
+	}
+
+	// parseStatements extracts statements from a sql file so we can execute
+	// them as separate migrations. OpenFGA uses Goose as their migration egine
+	// and Goose uses a single sql file for both up and down migrations.
+	// Grafana only supports up migration so we strip out the down migration
+	// and parse each individual statement
+	parseStatements := func(data []byte) ([]string, error) {
+		scripts := strings.Split(strings.TrimPrefix(string(data), "-- +goose Up"), "-- +goose Down")
+		if len(scripts) != 2 {
+			return nil, errors.New("malformed migration file")
+		}
+
+		// We assume that up migrations are always before down migrations
+		parts := strings.SplitAfter(scripts[0], ";")
+		stmts := make([]string, 0, len(parts))
+		for _, p := range parts {
+			p = strings.TrimSpace(p)
+			if p != "" {
+				stmts = append(stmts, p)
+			}
+		}
+
+		return stmts, nil
+	}
+
+	formatName := func(name string) string {
+		// Each migration file start with XXX where X is a number.
+		// We remove that part and prefix each migration with "zanzana".
+		return strings.TrimSuffix("zanzana"+name[3:], ".sql")
```

Comment:
what's the benefit to removing the number here? Seems like it may help down the line but I'm not aware of all the places where that number would should up. Just in the migration tracking table managed by the sqlstore package?

Reply:
to clarify, I think prefixing the migration is absolutely necessary but leaving the number seems acceptable

Reply:
It was done to make it more consistent with other grafana migrations. We could still keep them